workflows/prognostic_c48_run/runtime/monitor.py

```diff
@@ -0,0 +1,107 @@
+import logging
+from typing import (
+    Callable,
+    Iterable,
+    Set,
+)
+import vcm
+from runtime.types import Diagnostics, State
+
+from .names import DELP
+
+logger = logging.getLogger(__name__)
+
+
+class Monitor:
+    """Utility class for monitoring changes to a state dictionary and returning
+    the outputs as tendencies
+    """
+
+    def __init__(self, variables: Iterable[str], state: State, timestep: float):
+        """
+
+        Args:
+            variables: list of variables with names like
+                `tendency_of_{variable}_due_to_{name}`. Used to infer the variables
+                to be monitored.
+            state: The mutable object to monitor
+            timestep: the length of the timestep used to compute the tendency
+        """
+        # need to consume variables into set to use more than once
+        var_set = set(variables)
+        self.tendency_variables = filter_tendency(var_set)
+        self.storage_variables = filter_storage(var_set)
+        self._state = state
+        self.timestep = timestep
+
+    def __call__(
+        self, name: str, func: Callable[[], Diagnostics],
+    ) -> Callable[[], Diagnostics]:
+        """Decorator to add tendency monitoring to an update function
+
+        This will add the following diagnostics:
+        - `tendency_of_{variable}_due_to_{name}`
+        - `storage_of_{variable}_path_due_to_{name}`. A pressure-integrated version
+        of the above
+        - `storage_of_mass_due_to_{name}`, the total mass tendency in Pa/s.
+
+        Args:
+            name: the name to tag the tendency diagnostics with
+            func: a stepping function
+        Returns:
+            monitored function. Same as func, but with tendency and mass change
+            diagnostics inserted in place
+        """
+
+        def step() -> Diagnostics:
+
+            vars_ = self.storage_variables | self.tendency_variables
+            delp_before = self._state[DELP]
+            before = {key: self._state[key] for key in vars_}
+            diags = func()
+            delp_after = self._state[DELP]
```

Comment:
I think it works. If it didn't the mass change variable would always be zero, and it is not.

Reply:
It works for our particular code, but it's unsafe. This only works if the state dictionary is updated. It doesn't work if func updates the array inside the state dictionary instead of updating the state dictionary with a new array.

Reply:
I see your point, but this seems like a bigger issue about whether `DerivedFV3State.__getitem__` should exist and is not specific to this code. In practice, we treat `DataArray` as immutable objects in this code.